Add support for `pure_callback` and `debug.callback`

erick-xanadu opened the pull request:
**Context:** Similar to JAX, we would like to allow users a way to execute python code. We use the same names for denoting whether a function will return something or it is expected to not return anything and perform a side effect (like logging to a file, printing).

**Description of the Change:** Adds definitions for `pure_callback` and `debug.callback`.

[sc-61612]: Pure Callbacks
> **Requirements:**
> - A new function is added to Catalyst, `catalyst.pure_callback(callback_fn, result_type)(*args, **kwargs)`. Here:
>   - `callback_fn` is assumed to be a pure function
>   - `result_type` represents the result shape/dtype of the callback function, so that the callback can be compiled. In JAX, this is done by specifying `jax.ShapeDtypeStruct(return_shape, return_dtype)` for example.
>   - Using a callback should cause gradients to no longer work, since we cannot track the gradient of the callback.
>   - The docstring for `catalyst.callback()` should document any edge cases or restrictions.
> - The function should work both as a decorator and inline.
> - Arguments to the callback function should support any JAX jit compatible type (float, array, bool, dict, tuple, etc.)
> - Return values from the callback function should support any JAX jit compatible type (float, array, bool, dict, tuple, etc.)
> - The docstring for `pure_callback` should appear in the Sphinx documentation.

[sc-61613]: IO Callbacks
> **Requirements:**
> - A new function is added to Catalyst, `debug.callback(callback_fn)(*args, **kwargs)`. This function allows users to execute Python function (with no return values) _at runtime_ within their qjitted workflows. Since the users function does not permit any return values, this makes it useful for debugging.
> - Arguments to the callback function should support any JAX jit compatible type (float, array, bool, dict, tuple, etc.)
> - The docstring for `debug.callback` should appear in the Sphinx documentation.

[sc-60495]

Notes: documentation and position of `catalyst.debug.callback` is fixed in #621

erick-xanadu commented:
@josh146 , in JAX, the `io_callback` can return parameters. Do you want us to follow suit?

rmoyard commented on frontend/test/pytest/test_callback.py:
```diff
@@ -201,3 +201,102 @@ def cir(x):
+    assert captured.out.strip() == ""
+
+    with pytest.raises(ValueError, match="io_callback is expected to return None"):
+        cir(0)
```
```suggestion
        cir(0)

if __name__ == "__main__":
    pytest.main(["-x", __file__])

```

rmoyard commented on frontend/test/pytest/test_types.py:
```diff
@@ -56,3 +62,18 @@ def test_convert_numpy_dtype_to_mlir_error():
+)
+def test_convert_pytype_to_shaped_array(inp, exp):
+    """Test conversion to shaped arrays"""
+    assert convert_pytype_to_shaped_array(inp) == exp
```
```suggestion
    assert convert_pytype_to_shaped_array(inp) == exp

if __name__ == "__main__":
    pytest.main(["-x", __file__])
```

erick-xanadu commented on frontend/catalyst/pennylane_extensions.py:
```diff
@@ -2601,6 +2602,78 @@ def getArgsAsJAXArrays(self, flat_args):
+
+    result_type = tree_map(convert_pytype_to_shaped_array, result_type)
+    if result_type is None:
+        raise TypeError("pure_callback requires a result_type")
```
I don't think this is the same. The user can return something but we still need the result type to be explicit. How about the following compromise:

```suggestion
        raise TypeError("A function using pure_callback requires return types to be passed in as a parameter or type annotation.")

rmoyard replied:
Thanks I understand better :+1:

erick-xanadu commented on frontend/catalyst/pennylane_extensions.py:
```diff
@@ -2601,6 +2602,78 @@ def getArgsAsJAXArrays(self, flat_args):
+def io_callback(callback_fn):
+    """IO callback
+
+    Using `io_callback` allows a user to run python.
+    `io_callback`s can be used via a decorator:
```
How about the following?

```
    Using `io_callback` allows a user to run a python function with side effects inside an `@qjit` context.
    To mark a function as an `io_callback`, one can use a decorator:
```

I don't like the word "defined" here, because it usually means that the user writes the contents of `io_callback`. E.g.,

```python
def(ine) io_callback:
```

rmoyard replied:
Great!

dime10 commented on frontend/catalyst/pennylane_extensions.py:
````diff
@@ -2601,6 +2602,117 @@ def getArgsAsJAXArrays(self, flat_args):
+
+
+    ```python
+    def add_1(x) -> int:
+        return x + 1
+
+    @qjit
+    def context(x):
+        return pure_callback(add_1, int)(x)
````
Showcase the distinction since the type is passed explicitly?
```suggestion
    def add_1(x):
        return x + 1

    @qjit
    def context(x):
        return pure_callback(add_1, int)(x)
```

erick-xanadu commented on frontend/test/pytest/test_callback.py:
```diff
@@ -201,3 +201,102 @@ def cir(x):
+    """Test identity pure callback."""
+
+    def identity(a):
+        return a
```
I added a test to increment a variable. E.g.,

```
def inc(a):
  return a + 1
```

erick-xanadu commented:
@rmoyard, I've added the changes in a single commit. When you have time, please review. Thanks! :)

dime10 commented:
> @josh146 , in JAX, the `io_callback` can return return values. Do you want us to follow suit?

I was looking at the [JAX page for callbacks](https://jax.readthedocs.io/en/latest/notebooks/external_callbacks.html) again and I think we actually want `io_callback` to be `debug_callback`. An IO callback has the following restrictions:
- it does allow side-effects (printing, writing to disk, mutating Python state)
- it does allow return values (different from the current requirements/implementation)
- it is **not** AD compatible, unless all arguments do not participate in differentiation

The debug callback on the other hand:
- does not allow return values
- **is** AD compatible
- does support side-effects

The only "downside" I can see is that the debug callback (like the pure callback) is not guaranteed to execute (because it is intended to reflect what happens under the hood), while the IO callback is guaranteed to execute. Although it is our choice to implement this behaviour or not.

dime10 commented on frontend/test/pytest/test_callback.py:
```diff
@@ -201,3 +215,108 @@ def cir(x):
+    assert captured.out.strip() == "0"
+
+
+def test_io_callback_returns_something(capsys):
```
We might want to add a test for a different type of side-effect as well, like mutating a variable outside of its scope?

erick-xanadu commented:
> is AD compatible

How can something by AD compatible if it returns nothing? Won't it just be optimized away?

josh146 commented:
Hey @erick-xanadu @dime10; nice catch. Let's rename `io_callback` to `catalyst.debug.callback` (I'll update the ticket). However, we can do this in a follow up PR to avoid blocking this PR in the meantime.

> EDIT: I see now. One thing to note then is that in JAX debug.callback is only preserved for the forward pass and not the reverse pass. (Or at least, that's what it looks like in the example, as the print function is just called once).

Yep, I think this is a good approach; rather than registering `io_callback` with an 'identity' VJP transform, which is what we have in the ticket/requirements, I think this simplifies things. We don't need to register it with Enzyme, simply remove it from the backwards pass completely for the same effect.

erick-xanadu commented:
> Hey @erick-xanadu @dime10; nice catch. Let's rename io_callback to catalyst.debug.callback (I'll update the ticket). However, we can do this in a follow up PR to avoid blocking this PR in the meantime.

I'll just change the name, no worries. Thanks @josh146

erick-xanadu commented:
@josh146 @dime10 , I have updated the name and added a debug module inside callbacks. I will change the path to be `catalyst.debug.callback` in the following PR which addresses the print callback.

dime10 commented on frontend/catalyst/callback/debug/__init__.py:
Instead of adding a debug sub-module here, you can also import your callback function in the `catalyst/debug.py` module to achieve the desired import path `catalyst.debug.callback`. (this also avoids the issue of two debug modules existing if we were to try to import this one directly in `catalyst/__init__.py`)

dime10 commented on doc/code/__init__.rst:
```diff
@@ -27,3 +27,11 @@ Module: catalyst.cuda
     :no-inheritance-diagram:
     :fullname:
+
+Module: catalyst.callback
+---------------------
+
+.. automodapi:: catalyst.callback
+    :no-heading:
+    :no-inheritance-diagram:
+    :fullname:
```
Oh I see you want to this in a follow-up PR. In that case I would still remove the new doc module though.

erick-xanadu replied:
I removed it here: https://github.com/PennyLaneAI/catalyst/pull/661/commits/8a76c136e99da42355d755681ffaddeb298c4cf5

I will fix the path in the following PR.

dime10 reviewed (approved):
Thanks Erick!
